### ISSUE TITLE
perf(agent): bound production Qwen Q4_K prefill panels

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#9083]
Parent: #8972

## Current state

The exact production `fak serve` chat path selects `agent.InKernelPlanner`, not `modelengine.NativeScheduler`. On a cold Qwen3.5/3.8 resident Q4_K request, `generateReusedContextWithBias` calls `Session.Prefill` once for the entire divergent suffix. The frozen 32,800-token request therefore holds multiple full-P panels through each hybrid layer. macOS killed the exact fak listener at 58,329 MB compressed residency for `vm-compressor-space-shortage` during prefill before decode. #9066 is now closed and proves that nonzero-position Q4_K hybrid chunks stay on the resident path with all state preserved.

## Value

- For: operators serving long-context fak-native Qwen Q4_K requests through the ordinary production chat endpoint.
- Problem: one monolithic planner prefill scales temporary panels with the full prompt and can exhaust the host compressor before a token is decoded.
- Today: shrink the frozen request, call a different runtime, or use an unrelated modelengine scheduler path; none changes the failing production call chain.
- Better because: the same InKernelPlanner session can ingest bounded nonfinal chunks with PrefillNoLogits and obtain logits only from the final chunk, preserving native ownership and capping P-wide panel residency.
- Witness: a deterministic planner-level seam proves a long suffix is partitioned into bounded chunks on one session, every nonfinal chunk uses no-logits prefill, the final chunk alone returns logits, small/cached paths remain unchanged, and output/state matches monolithic prefill under the established Qwen oracle.

## Problem frame

- Centrality: Enabling; this is the exact production path required by #8972.
- P1: advanced; temporary prompt-panel residency is bounded at the real planner seam.
- P2: preserved; no phase time or memory gain is claimed until the live campaign reruns.
- P3: preserved; execution remains fak-native Q4_K on one session with no external fallback.
- P4: advanced; chunk ceiling and final-logit ownership become deterministic and testable.

## In scope

In `agent.InKernelPlanner` prefill, recognize the supported Qwen hybrid resident Q4_K envelope and split only the divergent prompt suffix into a fixed bounded chunk size. Call `Session.PrefillNoLogits` for every nonfinal chunk and `Session.Prefill` for the final chunk, preserve the existing single-session cache/state order and full-prompt snapshot timing, and keep the historical single-call path for other models/envelopes and small suffixes. Add focused deterministic partition/call-order/parity/cancellation tests.

## Out of scope

`modelengine.NativeScheduler`, inter-request prefill/decode interleaving, Metal kernels, Qwen math, admission policy, prefix-cache redesign, configurable tuning surfaces, throughput claims, llama.cpp, or #8972 artifact edits.

## Done condition

The production InKernelPlanner path bounds every supported Qwen hybrid resident Q4_K prefill call to at most 512 tokens, uses PrefillNoLogits for all nonfinal chunks and Prefill for the final chunk, preserves final logits and persistent state against monolithic prefill, checks cancellation between chunks, keeps non-Qwen/non-Q4_K and suffixes at or below 512 byte-identical, and passes exact-path validation.

## Lane

internal/agent native prefill

## Suggested paths

- internal/agent/inkernel_decode.go
- internal/agent/inkernel_prefill_chunk_test.go

## Acceptance

fak validate --mine internal/agent/inkernel_decode.go internal/agent/inkernel_prefill_chunk_test.go

## Closure

One signed commit with Fixes metadata and a recognized fak agent leaf trailer.

## Parent context

#8972

## Why this is next

#9066 makes resident Q4_K chunks append-safe. The failed endpoint never invokes #1912 NativeScheduler; it calls InKernelPlanner then one monolithic Session.Prefill. This leaf inserts the bound at that exact production seam.

## Working spine

Divergent suffix -> fixed 512-token partition -> same Session PrefillNoLogits for nonfinal chunks -> final Session.Prefill returns logits -> existing full-prompt snapshot/decode.

## Outcome

A 32,800-token production Qwen request cannot allocate planner-triggered Q4_K prefill panels wider than 512 tokens at a time.

## Dependencies

Builds on closed #9066. Blocks #8972. #1912 remains separate scheduler work and is not required by this leaf.

## Work unit

leaf

## Scale

S1

## Expected steps

4

## Assumptions

#9066 resident append parity is authoritative; `PrefillNoLogits` advances KV, convolution, recurrent state, and positions identically while discarding only final logits.

## Coordination notes

Own only the two named internal/agent paths. Coordinate with future #9075 work but do not edit its decode-coalescing surface. Do not edit internal/model or modelengine.

## Trigger

The exact clean #8972 production request passed tokenization/admission, ran 160.6416 seconds, then macOS memorystatus killed its fak PID for compressor exhaustion during monolithic prefill before decode.

## Batch policy

One agent-lane worker; deterministic/model-free tests first, then the existing #8972 hardware campaign is the live operating-envelope witness.

## Witness

- go test ./internal/agent -run TestInKernelQwenQ4KBoundedPrefill -count=1

## Acceptance gate

fak validate --mine internal/agent/inkernel_decode.go internal/agent/inkernel_prefill_chunk_test.go

## Likely files

- internal/agent/inkernel_decode.go
- internal/agent/inkernel_prefill_chunk_test.go

## Model tier

- Required model tier: tier/T2-required
- Optimal model tier: tier/T2-optimal

## Closure binding

Close only from one signed commit that fixes this issue, carries `(fak agent)`, proves the 512-token bound/call order/parity/cancellation and unchanged non-target paths, and passes exact-path validation.

**Comment by anthony-chaudhary** [comment]
Independent closure witness (2026-08-26 PT): `10ca2092eead4fcc815ac41f40c09f4c6a265c5f` is an ancestor of current origin/main and versions the changed surface as `internal/agent@r343+g10ca2092e`. The exact focused family `go test ./internal/agent -run '^TestInKernelQwenQ4KBoundedPrefill' -count=1` is green, and the full package `go test ./internal/agent -count=1` is green in 80.534s. `dos commit-audit` returns `OK`, `code_effect`, `diff-witnessed` over `inkernel_decode.go` plus its focused test. This witnesses the deterministic 512-token production planner seam only; the live #8972 Mac campaign remains the operating-envelope gate and no memory or parity gain is claimed here.
